### PR TITLE
UCT/DC: Add hash to track FC grant requests

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**
## What
Add hash to track FC grant requests

## Why ?
Fix flush race. After sending FC grant request and calling flush(CANCEL) we can't know whether request was sent and will it be answered. Here we implement logic which checks whether EP which initiated request still available.

### CODE REVIEW

File: src/uct/ib/dc/dc_mlx5.c [hunk 1]
```diff
@@ -871,7 +866,7 @@ static inline ucs_status_t uct_dc_mlx5_iface_flush_dcis(uct_dc_mlx5_iface_t *ifa
 {
     int i;
 
-    if (iface->tx.fc_grants) {
+    if (kh_size(&iface->tx.fc_hash)) {
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:869]: != 0

File: src/uct/ib/dc/dc_mlx5.c [hunk 2]
```diff
@@ -1014,21 +1011,24 @@ ucs_status_t uct_dc_mlx5_iface_fc_handler(uct_rc_iface_t *rc_iface, unsigned qp_
...
     } else if (fc_hdr == UCT_RC_EP_FC_PURE_GRANT) {
-        ep = *((uct_dc_mlx5_ep_t**)(hdr + 1));
+        sender = (uct_dc_fc_sender_data_t *)(hdr + 1);
 
-        if (!(ep->flags & UCT_DC_MLX5_EP_FLAG_VALID)) {
-            /* Just remove ep now, no need to clear waiting for grant state
-             * (it was done in destroy_ep func) */
-            uct_dc_mlx5_ep_release(ep);
+        it = kh_get(uct_dc_mlx5_fc_hash, &iface->tx.fc_hash, sender->ep);
+        if (it == kh_end(&iface->tx.fc_hash) ||
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:1017]: ()

File: src/uct/ib/dc/dc_mlx5.c [hunk 3]
```diff
@@ -1014,21 +1011,24 @@ ucs_status_t uct_dc_mlx5_iface_fc_handler(uct_rc_iface_t *rc_iface, unsigned qp_
...
+            kh_value(&iface->tx.fc_hash, it) != sender->payload.seq) {
+            /* Just ignore, ep was removed. We either not expecting grant on
+             * this EP or expecting another one */
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:1020]: .. or this is not the same grant sequence number we are expecting

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 4]
```diff
@@ -1072,17 +1049,25 @@ static UCS_CLASS_CLEANUP_FUNC(uct_dc_mlx5_ep_t)
...
+    uct_rc_txqp_purge_outstanding(&iface->super.super,
+                                  &iface->tx.dcis[self->dci].txqp, UCS_ERR_CANCELED,
+                                  iface->tx.dcis[self->dci].txwq.sw_pi, 1);
     ucs_assert(ucs_queue_is_empty(&iface->tx.dcis[self->dci].txqp.outstanding));
     iface->tx.dcis[self->dci].ep = NULL;
+
+    it = kh_get(uct_dc_mlx5_fc_hash, &iface->tx.fc_hash, (uint64_t)self);
+    if (it != kh_end(&iface->tx.fc_hash)) {
+        kh_del(uct_dc_mlx5_fc_hash, &iface->tx.fc_hash, it);
+        uct_dc_mlx5_ep_clear_fc_grant_flag(iface, self);
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c:1062]: can we remove the grant flag altogether?

Comment (Artemy-Mellanox) [reply src/uct/ib/dc/dc_mlx5_ep.c:1062]: AFAIU than we won't know we've already sent request for the EP and will keep sending

Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5_ep.c]: can we check in hash table?

Comment (Artemy-Mellanox) [reply src/uct/ib/dc/dc_mlx5_ep.c]: We can, but couldn't that affect performance?

Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5_ep.c]: is it done only when we want to send grant, and already do kh_put to add to table? maybe check kh_put result to see if key already exists?

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 5]
```diff
@@ -1344,15 +1346,13 @@ ucs_status_t uct_dc_mlx5_ep_check_fc(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_
 
     if (iface->super.super.config.fc_enabled) {
         UCT_RC_CHECK_FC_WND(&ep->fc, ep->super.stats);
-        if ((ep->fc.fc_wnd == iface->super.super.config.fc_hard_thresh) &&
-            !uct_dc_mlx5_ep_fc_wait_for_grant(ep)) {
+        if ((ep->fc.fc_wnd == iface->super.super.config.fc_hard_thresh)) {
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c:1349]: extra ()

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 6]
```diff
@@ -955,15 +955,18 @@ ucs_status_t uct_dc_mlx5_ep_fc_ctrl(uct_ep_t *tl_ep, unsigned op,
...
         it = kh_put(uct_dc_mlx5_fc_hash, &iface->tx.fc_hash, sender.ep, &ret);
-        if (ret == UCS_KH_PUT_FAILED) {
+        if (ret == UCS_KH_PUT_KEY_PRESENT) {
+            return UCS_OK;
+        } else if (ret == UCS_KH_PUT_FAILED) {
             ucs_error("failed to create hash entry for fc hard req");
             return UCS_ERR_NO_MEMORY;
         }
+
+        sender.payload.seq               = iface->tx.fc_seq++;
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c:967]: for consistency, can set sender.ep here, and pass (uintptr)ep to kh_put

**Review by yosefe — APPROVED** [review]
pls squash

File: src/uct/ib/dc/dc_mlx5_ep.h [hunk 7]
```diff
@@ -23,23 +23,17 @@ enum uct_dc_mlx5_ep_flags {
...
-    UCT_DC_MLX5_EP_FLAG_KEEPALIVE_POSTED    = UCS_BIT(4),
+    UCT_DC_MLX5_EP_FLAG_KEEPALIVE_POSTED    = UCS_BIT(3),
 
     /* Flush cancel was executed on EP */
-    UCT_DC_MLX5_EP_FLAG_FLUSH_CANCEL        = UCS_BIT(5),
+    UCT_DC_MLX5_EP_FLAG_FLUSH_CANCEL        = UCS_BIT(4),
 
     /* Error handler already called or flush(CANCEL) disabled it */
-    UCT_DC_MLX5_EP_FLAG_ERR_HANDLER_INVOKED = UCS_BIT(6),
+    UCT_DC_MLX5_EP_FLAG_ERR_HANDLER_INVOKED = UCS_BIT(5),
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5_ep.h:36]: no need for trailing comma

**Review by brminich — APPROVED** [review]

File: src/uct/ib/dc/dc_mlx5_ep.h (hunk 7 above)
Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5_ep.h:36]: @Artemy-Mellanox can force push the fix

Comment (Artemy-Mellanox) [reply src/uct/ib/dc/dc_mlx5_ep.h:36]: @yosefe i thought we decided to keep it, are you sure?

Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5_ep.h:36]: can you pls remind what was reason? to avoid extra diff when adding new elements?

Comment (Artemy-Mellanox) [reply src/uct/ib/dc/dc_mlx5_ep.h:36]: yes and that we can't maintain compatibility with C89 anyway

Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5_ep.h:36]: ok

**Pull request merged**